pkg/storage/unified/resource/kv.go

```diff
@@ -0,0 +1,207 @@
+package resource
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"iter"
+	"time"
+
+	badger "github.com/dgraph-io/badger/v4"
+)
+
+var ErrNotFound = errors.New("key not found")
+
+type SortOrder int
+
+const (
+	SortOrderAsc SortOrder = iota
+	SortOrderDesc
+)
+
+type ListOptions struct {
+	Sort     SortOrder // sort order of the results. Default is SortOrderAsc.
+	StartKey string    // lower bound of the range, included in the results
+	EndKey   string    // upper bound of the range, excluded from the results
+	Limit    int64     // maximum number of results to return. 0 means no limit.
+}
+
+// KVObject represents a key-value object
+type KVObject struct {
+	Key   string        // the key of the object within the section
+	Value io.ReadCloser // the value of the object
+}
+
+type KV interface {
+	// Keys returns all the keys in the store
+	Keys(ctx context.Context, section string, opt ListOptions) iter.Seq2[string, error]
+
+	// Get retrieves a key-value pair from the store
+	Get(ctx context.Context, section string, key string) (KVObject, error)
+
+	// Save a new value
+	Save(ctx context.Context, section string, key string, value io.ReadCloser) error
```

Comment:
I would suggest using `io.Reader` only, to leave responsibility to close to the caller. Caller knows it can close after `Save` returns.